### PR TITLE
Yulproto funcdef stmt

### LINKED ISSUE #7276: [yul] proto fuzzer: Make function definition a statement
Move from top-level function definitions to those that can be declared anywhere as long as yul rules are respected.

### PR DESCRIPTION
**Pull request by bshastry**
fixes #7276

### CODE REVIEW

**Comment by chriseth** [comment]
please rebase.

**Comment by bshastry** [comment]
Rebased PR on top of target branch. Once the review passes, please merge this onto develop after #7309

File: test/tools/ossfuzz/protoToYul.h
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.h]: Is the difference in indentation on purpose?

File: test/tools/ossfuzz/protoToYul.cpp
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]: Can't this be combined with the above case?

File: test/tools/ossfuzz/protoToYul.cpp
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]:
```suggestion
		if (m_functionSigMap.size() > 0)
		{
```

File: test/tools/ossfuzz/protoToYul.cpp
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]:
```suggestion
				switch (numOutParams)
				{
```

File: test/tools/ossfuzz/protoToYul.cpp
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]:
```suggestion
void ProtoConverter::updateFunctionMaps(string const& var)
```

File: test/tools/ossfuzz/protoToYul.cpp
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]: Is this still needed?

Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp]: sorry, I meant to refactor the commented out portion. Will push an update.

**Comment by bshastry** [comment]
Updated PR.

**Comment by chriseth** [comment]
I think this PR is rather important to get in quickly, but I currently do not have the resources to look at it in depth. I'm fine to merge this if it generates valid and meaningful code.

**Comment by bshastry** [comment]
@chriseth I have been fuzzing this PR over the last couple of days locally and didn't find anything unexpected. For reference, here are some programs generated by it.

https://gist.github.com/bshastry/dde67b9239cef4ab71f953184a731705

I am currently investigating why test coverage slightly drops with this PR but I feel in the long run, this might not be a problem. Will update shortly with a more conclusive opinion.

**Comment by chriseth** [comment]
Ok, great!

Do you know why most of the programs are rather "uniform" - example: https://gist.github.com/bshastry/dde67b9239cef4ab71f953184a731705#file-100_yul_programs_generated_by_pr7318-L2196

**Review by chriseth — APPROVED** [review]

**Pull request merged**
